torchdata/datapipes/iter/transform/callable.py

```diff
@@ -168,6 +170,155 @@ def __len__(self) -> int:
+        >>> fully_shuffled_flatmapped_dp = source_dp.in_batch_shuffle()
+        >>> fully_shuffled_flatmapped_dp = fully_shuffled_flatmapped_dp.shuffled_flatmap()
+        >>> list(fully_shuffled_flatmapped_dp)
+        ['b', 3, 'c', 'd', 'C', 'A', 'a', 2, 'B', 'D', 4, 1]
+    """
+    datapipe: IterDataPipe[T_co]
+    fn: Optional[Callable]
+    buffer_size: int
+    _buffer: List[Iterator[T_co]]
+    _enabled: bool
+    _seed: Optional[int]
+    _rng: random.Random
+
+    def __init__(
+        self, datapipe: IterDataPipe, fn: Optional[Callable] = None, input_col=None, buffer_size: int = 100
+    ) -> None:
+        super().__init__()
+        self._buffer = []
+        self.datapipe = datapipe
+
+        if fn is None:
+            fn = _no_op_fn
+        _check_unpickable_fn(fn)
+        self.fn = fn  # type: ignore[assignment]
+        self.input_col = input_col
+        validate_input_col(fn, input_col)
+
+        assert buffer_size > 0, "buffer_size should be larger than 0"
+        self.buffer_size = buffer_size
+        self._enabled = True
+        self._seed = None
+        self._rng = random.Random()
+
+    def set_shuffle(self, shuffle=True):
+        self._enabled = shuffle
+        return self
+
+    def set_seed(self, seed: int):
+        self._seed = seed
+        return self
+
+    def reset(self) -> None:
+        self._buffer = []
+        if self._enabled:
+            if self._seed is None:
+                self._seed = int(torch.empty((), dtype=torch.int64).random_().item())
+            self._rng.seed(self._seed)
+            self._seed = None
+
+    def _apply_fn(self, data):
+        if self.input_col is None:
+            return self.fn(data)  # type: ignore[misc]
+        elif isinstance(self.input_col, (list, tuple)):
+            args = tuple(data[col] for col in self.input_col)
+            return self.fn(*args)  # type: ignore[misc]
+        else:
+            return self.fn(data[self.input_col])  # type: ignore[misc]
+
+    def __iter__(self) -> Iterator[T_co]:
+        if not self._enabled:  # equivalent to flatmap
+            for x in self.datapipe:
+                yield from self._apply_fn(x)
+        else:
+            idx = self._rng.randint(0, self.buffer_size - 1)
+            for x in self.datapipe:
+                while len(self._buffer) == self.buffer_size:
+                    try:
+                        yield next(self._buffer[idx])
+                        idx = self._rng.randint(0, self.buffer_size - 1)
+                    except StopIteration:
+                        self._buffer.pop(idx)
+                self._buffer.append(iter(self._apply_fn(x)))
+            while self._buffer:
+                try:
+                    idx = self._rng.randint(0, len(self._buffer) - 1)
+                    yield next(self._buffer[idx])
+                except StopIteration:
+                    self._buffer.pop(idx)
```

Comment:
We can simplify this part by directly using the result from `self._apply_fn(x)` rather than `iter`. We don;t need to handle `StopIteration` as we will be able to know the size of output.
WDYT?

Reply:
I think it is nice to allow the `fn` to return iterables and not just sequences. This datapipe will be used only in very specific cases, otherwise, it can simply be replaced by `dp.flatmap().shuffle()`. So I would keep this freedom

Reply:
Then, why not do `list` for iterables so that no need to handle `StopIteration`

Reply:
If I'm not mistaken, a call to `list` would load all the items in the iterable, right?
Then, we would lose the whole interest of this datapipe, and it would be simpler to just use `dp.flatmap().shuffle()`.

Reply:
> If I'm not mistaken, a call to `list` would load all the items in the iterable, right?

Yeah. And, it would only load the output of each flatmap function, right?



> Then, we would lose the whole interest of this datapipe, and it would be simpler to just use `dp.flatmap().shuffle()`.

First, it's a syntax sugar, right? Second, this DataPipe provides more randomness as `dp.flatmap().shuffle()` would do buffered shuffle over the individual output from flatmap, and this DataPipe performs shuffle at the source.

Reply:
Actually, if we want to keep `randint` being called once per iteration, it should be doable by mapping the original `idx` from `[0, len(self._buffer) - 1]` to `[0, len(self._buffer) - 2]`

Reply:
> Actually, if we want to keep `randint` being called once per iteration, it should be doable by mapping the original `idx` from `[0, len(self._buffer) - 1]` to `[0, len(self._buffer) - 2]`

Is it fine if the sampling is not uniform? (for example with this mapping:`idx % len(self._buffer)`)

Reply:
About `list` vs `iter`, here s a concrete example:
```python
from torch.utils.data import MapDataPipe
from torch.utils.data.datapipes.iter import FileLister
from torch.utils.data.datapipes.map import Shuffler

import mne


def read_epochs(filename) -> mne.Epochs:
    return mne.read_epochs(filename, preload=False)


class EpochsWrapper(MapDataPipe):
    def __init__(self, epochs: mne.Epochs):
        super().__init__()
        self.epochs = epochs

    def __len__(self):
        return len(self.epochs)

    def __getitem__(self, item):
        return self.epochs.get_data(item=item)


# ~/data contains 100 files of 1GB each
root = '~/data/'

datapipe = FileLister(root) \
    .map(read_epochs) \
    .in_memory_cache() \
    .map(EpochsWrapper) \
    .map(Shuffler) \
    .shuffled_flatmap()
```
This would be similar to typical pipelines I would encounter.
In this example, we don't want to load in memory all the examples in each file because this would be 100GB.

If we used `list` instead of `iter`, the 100GB would be loaded in memory when filling the buffer right? (I am honestly unsure about this affirmation, this is not a rhetorical question, please help me if I missed something)

Or do you have another idea to shuffle this kind of data?

Reply:
Okay, I see. It makes sense to keep it as a generator to prevent loading extra data in memory ahead of time within the buffer. 

BTW, I am not sure your pipeline is correct, though. `EpochsWrapper` is not a callable to be sent to `dp.map`.

So, let's keep it as `iter` to preserve the benefit of lazy loading within `flatmap_fn`.

> Is it fine if the sampling is not uniform? (for example with this mapping:`idx % len(self._buffer)`)

That's actually a good argument not doing that.

Emmm. Then, let's keep it as it is. It should be good as long as the operations are deterministic